benchtop/src/workload.rs

```diff
@@ -23,66 +24,155 @@ pub trait Workload: Send {
+                ),
+            ),
         ),
         "randw" => (
             Box::new(custom_workload::init(db_size)),
-            dyn_vec(custom_workload::build(
-                0,
-                100,
-                workload_size,
-                fresh.unwrap_or(0),
-                db_size,
-                op_limit,
+            dyn_vec(
+                cache_size,
                 threads,
-            )),
+                custom_workload::build(
+                    0,
+                    100,
+                    workload_size,
+                    fresh.unwrap_or(0),
+                    db_size,
+                    op_limit,
+                    threads as usize,
+                ),
+            ),
         ),
         "randr" => (
             Box::new(custom_workload::init(db_size)),
-            dyn_vec(custom_workload::build(
-                100,
-                0,
-                workload_size,
-                fresh.unwrap_or(0),
-                db_size,
-                op_limit,
+            dyn_vec(
+                cache_size,
                 threads,
-            )),
+                custom_workload::build(
+                    100,
+                    0,
+                    workload_size,
+                    fresh.unwrap_or(0),
+                    db_size,
+                    op_limit,
+                    threads as usize,
+                ),
+            ),
         ),
         "randrw" => (
             Box::new(custom_workload::init(db_size)),
-            dyn_vec(custom_workload::build(
-                50,
-                50,
-                workload_size,
-                fresh.unwrap_or(0),
-                db_size,
-                op_limit,
+            dyn_vec(
+                cache_size,
                 threads,
-            )),
+                custom_workload::build(
+                    50,
+                    50,
+                    workload_size,
+                    fresh.unwrap_or(0),
+                    db_size,
+                    op_limit,
+                    threads as usize,
+                ),
+            ),
         ),
         name => anyhow::bail!("invalid workload name: {}", name),
     })
 }
+
+struct LruCacheWorkload<W> {
+    cache: LruCache<Vec<u8>, Option<Vec<u8>>>,
+    inner: W,
+}
+
+impl<W: Workload> LruCacheWorkload<W> {
+    fn new(inner: W, cache_size: usize) -> Self {
+        LruCacheWorkload {
+            inner,
+            cache: LruCache::new(cache_size.try_into().expect("non-zero cache size")),
+        }
+    }
+}
+
+impl<W: Workload> Workload for LruCacheWorkload<W> {
+    fn run_step(&mut self, transaction: &mut dyn Transaction) {
+        let mut tx = LruCacheTransaction {
+            inner: transaction,
+            cache: &mut self.cache,
+        };
+        self.inner.run_step(&mut tx);
+    }
+
+    fn is_done(&self) -> bool {
+        self.inner.is_done()
+    }
+}
+
+struct LruCacheTransaction<'a> {
```

Comment:
Aren't the lifetimes for `Transaction` and `LruCache` different? i.e. `LruCache` outlives the `Transaction. Do you use the same one for simplicity?

Reply:
I didn't write the code, so I could be wrong about the intention, but the main thing that the lifetime should provide to the compiler is that all references present in `LruCacheTransaction` must outlive the struct itself, and given its usage, they do. It doesn't really matter if `Transaction` of `LruCache` lives longer compared to the other, so I would say that for simplicity, they take the same lifetime

Reply:
Yes, it's for simplicity!

What's important isn't the actual lifetimes of the objects, but their relationships to each other.